packages/falso/src/lib/core/core.ts

```diff
@@ -1,40 +1,116 @@
 import { random } from '../random';
+import { primitiveValueIsUnique } from './unique-validators';
 
 export interface FakeOptions {
   length?: number;
+  priority?: 'length' | 'unique';
   locale?: any | string[];
 }
 
 export type markRequired<Type, Key extends keyof Type> = Type & {
   [Property in Key]-?: Type[Property];
 };
 
+export interface FakeConfig<T> {
+  uniqueComparer: (
+    item: T,
+    items: T[],
+    comparisonKeys?: (keyof T)[]
+  ) => boolean | ((item: T, items: T[]) => boolean);
+  comparisonKeys?: (keyof T)[];
+}
+
 export type Return<T, O extends FakeOptions> = [O] extends [never]
   ? T
   : O['length'] extends number
   ? T[]
   : T;
 
-type FactoryFunction<T> = (i: number) => T;
+export type FactoryFunction<T> = (i: number) => T;
 
 export function fake<T, Options extends FakeOptions>(
   data: Readonly<T[]> | FactoryFunction<T>,
-  options?: Options
+  options?: Options,
+  config: FakeConfig<T> = {
+    uniqueComparer: primitiveValueIsUnique,
+  }
 ): Return<T, Options> {
-  const dataSource = Array.isArray(data)
-    ? () => randElement(data)
-    : (data as FactoryFunction<T>);
+  if (options?.length === 0) {
+    return [] as any;
+  }
+
+  if (Array.isArray(data)) {
+    return fakeFromArray(data, options) as any;
+  }
+
+  return fakeFromFunction(data as FactoryFunction<T>, config, options) as any;
```

Comment:
You have a lot of shared code here, I think I would normalize the array data source to also work on a function with a random element and just use factory-based fake for both.

Reply:
I'm happy to do this I just felt like this whole `fake` logic was getting quite complex so keeping them separate made it easier to read for me. But i'm happy either way.

I think in hindsight, if we were starting over i'd suggest that instead of `option.length` property we have singular and plural function for each. E.G `randId` that returns a `string` & `randIds` that uses `randId` to generate & return a `string[]`. But it might be a bit late for that ;) haha

Reply:
Personally I prefer the length property, adding a plural will double the API for not a good enough reason from my point of view

Reply:
Fair play. It works pretty well for our users, it just adds a lot of headaches for deving haha. All good though